qiime2/sdk/action.py

```diff
@@ -284,6 +282,47 @@ def _build_annotations(self):
 
         return annotations
 
+    def _build_numpydoc(self):
+        numpydoc = ""
+        numpydoc += "{}\n\n".format(self.name)
+        numpydoc += "{}\n\n".format(self.description)
+
+        sig = self.signature
+        params = True if len(sig.inputs) != 0 else False
+
+        numpydoc += "{}".format(self._build_section(
+            "Parameters", sig.inputs))
+        numpydoc += "{}".format(self._build_section(
+            "Parameters", sig.parameters, parameters=params))
+        numpydoc += ("\n" if len(sig.parameters) != 0 or
+                     len(sig.inputs) != 0 else "")
+        numpydoc += "{}\n".format(self._build_section(
+            "Returns", sig.outputs))
+
+        return numpydoc
+
+    def _build_section(self, header, iterable, parameters=False):
+        section = ""
+
+        if len(iterable) != 0:
+            if not parameters:
+                section += "{}\n".format(header)
+                section += "{}\n".format(''.join(['-' for char in header]))
+            for item in iterable:
+                item_obj = iterable[item]
+                section += "{}: {}".format(item, item_obj.qiime_type)
+                if type(item_obj.default).__name__ != "_NoValue":
```

Comment:
Calling `type(stuff).__anythinghere__` should be a hint that you are breaking someone's abstractions. Sometimes they are bad abstractions, or you are doing something very low level, but it should give you pause.